### PR TITLE
Fix serverless job run failed related issues

### PR DESCRIPTION
**Pull request by t-rufang**
With this PR,  
 - Serverless job can successfully run and finish when backend is stable.
 - Pop up Spark master UI web page when serverless job is starting
 - Warn user that kill serverless job action is invalid when the job is in `Finalizing` or `Ended` state

### CODE REVIEW

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 1]
```diff
@@ -260,94 +249,143 @@ public String getJobUuid() {
...
+    }
+
+    @Override
+    public boolean isSuccess(String state) {
+        return state.equalsIgnoreCase(SparkBatchJobState.SUCCESS.toString());
+    }
+
+    public boolean isJobEnded(
+            @NotNull com.microsoft.azure.hdinsight.sdk.rest.azure.serverless.spark.models.SparkBatchJob sparkBatchJob) {
+        return sparkBatchJob.schedulerState().toString().equalsIgnoreCase(SchedulerState.ENDED.toString());
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: Do you need adding some abnormal end status, such as: failed, killed?

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 2]
```diff
@@ -260,94 +249,143 @@ public String getJobUuid() {
...
-                                                .zipWith(Observable.range(1, getRetriesMax()), (n, i) -> i)
-                                                .doOnNext(i -> ctrlInfo("Retry retrieving livy submission log..."))
-                                                .delay(getDelaySeconds(), TimeUnit.SECONDS);
-                                    }
-                                }))
-                )
-                .onErrorResumeNext(err -> {
-                    if (err instanceof SparkJobFinishedException || err.getCause() instanceof SparkJobFinishedException) {
-                        return Observable.error(err);
+                .takeUntil(sparkBatchJob ->
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: The val `sparkBatchJob` might be easy confused. You can try `batchResponse` or `batchResp`

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 3]
```diff
@@ -166,85 +163,77 @@ public String getJobUuid() {
...
-                        new AbstractMap.SimpleImmutableEntry<>(
-                                responsePayload.getState(),
-                                String.join("\n", responsePayload.getLog())))
-                .onErrorResumeNext(err -> {
-                    if (err instanceof SparkJobFinishedException) {
-                        return Observable.error(err);
+        final int GET_JOB_DONE_REPEAT_DELAY_MILLISECONDS = 1000;
+        return getSparkBatchJobRequest()
+                .repeatWhen(ob -> ob.delay(GET_JOB_DONE_REPEAT_DELAY_MILLISECONDS, TimeUnit.MILLISECONDS))
+                .takeUntil(sparkBatchJob -> {
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: The same, `batchResp` or others

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 4]
```diff
@@ -166,85 +163,77 @@ public String getJobUuid() {
...
+                    String jobState = getJobState(sparkBatchJob);
+                    return isJobEnded(sparkBatchJob)
+                            || (jobState != null && isDone(jobState));
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: What's the difference between `isDone` and `isJobEnded`? Can they be merged?

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 5]
```diff
@@ -166,85 +163,77 @@ public String getJobUuid() {
...
-                        throw propagate(exception);
-                    } else {
-                        // Retry with limited times
-                        return Observable.just(1)
-                                .zipWith(Observable.range(1, getRetriesMax()), (n, i) -> i)
-                                .delay(getDelaySeconds(), TimeUnit.SECONDS);
+        ctrlInfo("The Spark job is starting...");
+        return getSparkBatchJobRequest()
+                .doOnNext(sparkBatchJob -> {
+                    if (sparkBatchJob.properties() != null && sparkBatchJob.properties().sparkMasterUI() != null) {
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: The `sparkBatchJob.properties() != null && sparkBatchJob.properties().sparkMasterUI()` can be wrapped as a `getMasterUI(batchResp)` method

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 6]
```diff
@@ -166,85 +163,77 @@ public String getJobUuid() {
...
-                    }
-                });
+                })
+                .flatMap(job ->
+                        getSparkBatchJobRequest()
+                                .repeatWhen(ob -> ob.delay(getDelaySeconds(), TimeUnit.SECONDS))
+                                .takeUntil(sparkBatchJob -> {
+                                    String jobState = getJobState(sparkBatchJob);
+                                    return isJobEnded(sparkBatchJob)
+                                            || (jobState != null && (isDone(jobState) || isRunning(jobState)));
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: If `isJobEnded` and `isDone` merged, the code can looks like: `isJobEnded() || isJobStarted()`

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 7]
```diff
@@ -166,85 +163,77 @@ public String getJobUuid() {
...
-                        return Observable.error(
-                                new SparkJobException(errorMsg));
-                    } else {
-                        return Observable.just(responsePayload.getState());
-                    }
-                });
+                })
+                .flatMap(job ->
+                        getSparkBatchJobRequest()
+                                .repeatWhen(ob -> ob.delay(getDelaySeconds(), TimeUnit.SECONDS))
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: You can keep `"The Spark job is starting..."` output in repeat waiting function.

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 8]
```diff
@@ -166,85 +163,77 @@ public String getJobUuid() {
...
+                                })
+                                .filter(sparkBatchJob -> {
+                                    String jobState = getJobState(sparkBatchJob);
+                                    return isJobEnded(sparkBatchJob)
+                                            || (jobState != null && (isDone(jobState) || isRunning(jobState)));
+                                })
+                                .flatMap(sparkBatchJob -> {
+                                    String jobState = getJobState(sparkBatchJob);
+                                    if (jobState != null && (isDone(jobState) || isRunning(jobState))) {
+                                        if (isDone(jobState) && !isSuccess(jobState)) {
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: Revert the `if..else` might be more clear and easy understanding.

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 9]
```diff
@@ -260,94 +249,143 @@ public String getJobUuid() {
...
-                                    }
-                                }))
-                )
-                .onErrorResumeNext(err -> {
-                    if (err instanceof SparkJobFinishedException || err.getCause() instanceof SparkJobFinishedException) {
-                        return Observable.error(err);
+                .takeUntil(sparkBatchJob ->
+                        isJobEnded(sparkBatchJob)
+                                || (sparkBatchJob.properties() != null
+                                && StringUtils.isNotEmpty(sparkBatchJob.properties().livyServerAPI())))
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: The same as above,  `getLivyAPI()`

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 10]
```diff
@@ -260,94 +249,143 @@ public String getJobUuid() {
...
+                                                .takeUntil(logAndStateTriple ->
+                                                        !logAndStateTriple.getMiddle().equalsIgnoreCase(SparkBatchJobState.STARTING.toString())
+                                                                || logAndStateTriple.getRight().equalsIgnoreCase(SchedulerState.ENDED.toString())
+                                                )
+                                                .flatMap(logAndStateTriple -> {
+                                                    if (!logAndStateTriple.getMiddle().equalsIgnoreCase(SparkBatchJobState.STARTING.toString())) {
+                                                        String message = "Job is in " + logAndStateTriple.getMiddle() + " state.";
+                                                        return Observable.just(new AbstractMap.SimpleImmutableEntry<>(Log, message));
+                                                    } else if (logAndStateTriple.getRight().equalsIgnoreCase(SchedulerState.ENDED.toString())) {
+                                                        String message = "Job is in " + logAndStateTriple.getRight() + " state.";
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: It's almost the same message than above job state, need to fix it. And a question, do we need to expose those states to users? If yes, should they be present both, such as: <jobState>-<schedulerState>?

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java (hunk 6 above)
Comment (t-rufang) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: Yes, these two methods are confusing, I will merge them then.

---
In reply to: [268020311](https://github.com/Microsoft/azure-tools-for-java/pull/2899#discussion_r268020311) [](ancestors = 268020311)

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 11]
```diff
@@ -260,94 +241,194 @@ public String getJobUuid() {
...
+        return getSparkBatchJobRequest()
+                .retryWhen(err ->
+                        err.zipWith(Observable.range(1, getRetriesMax()), (n, i) -> i)
+                                .delay(GET_LIVY_URL_RETRY_DELAY_MILLISECONDS, TimeUnit.MILLISECONDS)
                 )
-                .onErrorResumeNext(err -> {
-                    if (err instanceof SparkJobFinishedException || err.getCause() instanceof SparkJobFinishedException) {
-                        return Observable.error(err);
+                .repeatWhen(ob -> ob.delay(GET_LIVY_URL_REPEAT_DELAY_MILLISECONDS, TimeUnit.MILLISECONDS))
+                .takeUntil(batchResp -> isJobEnded(batchResp) || StringUtils.isNotEmpty(getLivyAPI(batchResp)))
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: How about if `getJobState()` always returns `null`?

Comment (t-rufang) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: Throw an exception in the new iteration.

---
In reply to: [268477836](https://github.com/Microsoft/azure-tools-for-java/pull/2899#discussion_r268477836) [](ancestors = 268477836)

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 12]
```diff
@@ -204,6 +214,11 @@ public String getJobUuid() {
                 })
                 .flatMap(job ->
                         getSparkBatchJobRequest()
+                                .flatMap(batchResp ->
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: May try to use one `getSparkBatchJobRequest()` here

**Pull request merged**